File: app/src/main/java/com/surya/shopcart/adapter/CartItemAdapter.java
package com.surya.shopcart.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.squareup.picasso.Picasso;
import com.surya.shopcart.Product;
import com.surya.shopcart.ProductAdapter;
import com.surya.shopcart.R;
import com.surya.shopcart.utils.Utils;

import java.util.ArrayList;

import okhttp3.internal.Util;

public class CartItemAdapter extends RecyclerView.Adapter<CartItemAdapter.ViewHolder> {

    private OnItemClickListener listener;

    private ArrayList<Product> cartItemsList;

    FirebaseStorage storage = FirebaseStorage.getInstance();

    public CartItemAdapter(ArrayList<Product> cartItemsList) {
        this.cartItemsList = cartItemsList;
    }

    public void setOnItemClickListener(CartItemAdapter.OnItemClickListener listener) {
        this.listener = listener;
    }
    @NonNull
    @Override
    public CartItemAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.cart_item_layout, parent, false);
        return new CartItemAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Product product = cartItemsList.get(position);

        StorageReference storageRef = storage.getReference();
        StorageReference pathReference = storageRef.child(product.getImage());

        pathReference.getDownloadUrl().addOnSuccessListener(uri -> {
            Picasso.get().load(uri).into(holder.image);
        }).addOnFailureListener(exception -> {
            Log.e("TAG", "Failed to load image from Firebase Storage", exception);
        });

        holder.title.setText(product.getTitle());
        holder.price.setText("$ "+product.getPrice() +" /lb");
        holder.quantity.setText(product.getQuantity()+"");
        holder.subTotal.setText("$ "+Utils.getSubtotalStr(product.getQuantity(), product.getPrice()) + " (Subtotal)");
    }

    @Override
    public int getItemCount() {
        return cartItemsList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder{
        ImageView image;
        TextView title;
        TextView quantity;
        TextView price;
        TextView reduceQuantity;
        TextView increaseQuantity;
        TextView subTotal;

        public ViewHolder(View view){
            super(view);
            image = itemView.findViewById(R.id.image);
            title = itemView.findViewById(R.id.title);
            price = itemView.findViewById(R.id.price);
            quantity = itemView.findViewById(R.id.quantity);
            reduceQuantity = itemView.findViewById(R.id.reduceQuantity);
            increaseQuantity = itemView.findViewById(R.id.increaseQuantity);
            subTotal = itemView.findViewById(R.id.subTotal);

            image.setOnClickListener(view1 -> {
                if (listener != null) {
                    int position = getAdapterPosition();
                    if (position != RecyclerView.NO_POSITION) {
                        listener.onItemClick(view1, position, "image");
                    }
                }
            });

            reduceQuantity.setOnClickListener(view1 -> {
                if (listener != null) {
                    int position = getAdapterPosition();
                    if (position != RecyclerView.NO_POSITION) {
                        listener.onItemClick(view1, position, "reduceQuantity");
                    }
                }
            });

            increaseQuantity.setOnClickListener(view1 -> {
                if (listener != null) {
                    int position = getAdapterPosition();
                    if (position != RecyclerView.NO_POSITION) {
                        listener.onItemClick(view1, position, "increaseQuantity");
                    }
                }
            });
        }
    }

    public interface OnItemClickListener {
        void onItemClick(View view, int position, String type);
    }
}

File: app/src/main/java/com/surya/shopngo/adapter/CartItemAdapter.kt
package com.surya.shopngo.adapter

import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.firebase.storage.FirebaseStorage
import com.squareup.picasso.Picasso
import com.surya.shopngo.dataclass.Product
import com.surya.shopngo.R
import com.surya.shopngo.utils.Utils

class CartItemAdapter : RecyclerView.Adapter<CartItemAdapter.ViewHolder> {
    private var listener: OnItemClickListener? = null
    private var cartItemsList: ArrayList<Product>
    var storage = FirebaseStorage.getInstance()
    var showQuantityLayout = true

    constructor(cartItemsList: ArrayList<Product>) {
        this.cartItemsList = cartItemsList
    }

    constructor(cartItemsList: ArrayList<Product>, showQuantityLayout: Boolean) {
        this.cartItemsList = cartItemsList
        this.showQuantityLayout = showQuantityLayout
    }

    fun setOnItemClickListener(listener: OnItemClickListener?) {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            LayoutInflater.from(parent.context).inflate(R.layout.cart_item_layout, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val product = cartItemsList[position]
        val storageRef = storage.getReference()
        val pathReference = storageRef.child(product.image)
        pathReference.getDownloadUrl()
            .addOnSuccessListener { uri: Uri? -> Picasso.get().load(uri).into(holder.image) }
            .addOnFailureListener { exception: Exception? ->
                Log.e(
                    "TAG",
                    "Failed to load image from Firebase Storage",
                    exception
                )
            }
        holder.title.text = product.title
        holder.price.text = "$ " + product.price + " /lb"
        holder.quantity.text = product.quantity.toString() + ""
        holder.subTotal.text =
            "$ " + Utils.getSubtotalStr(product.quantity, product.price) + " (Subtotal)"
        if (!showQuantityLayout) {
            holder.quantityLayout.visibility = View.GONE
        }
    }

    override fun getItemCount(): Int {
        return cartItemsList.size
    }

    inner class ViewHolder(view: View?) : RecyclerView.ViewHolder(view!!) {
        var image: ImageView
        var title: TextView
        var quantity: TextView
        var price: TextView
        var reduceQuantity: TextView
        var increaseQuantity: TextView
        var subTotal: TextView
        var quantityLayout: LinearLayout

        init {
            image = itemView.findViewById(R.id.image)
            title = itemView.findViewById(R.id.title)
            price = itemView.findViewById(R.id.price)
            quantity = itemView.findViewById(R.id.quantity)
            reduceQuantity = itemView.findViewById(R.id.reduceQuantity)
            increaseQuantity = itemView.findViewById(R.id.increaseQuantity)
            subTotal = itemView.findViewById(R.id.grandTotal)
            quantityLayout = itemView.findViewById(R.id.quantityLayout)
            image.setOnClickListener { view1: View? ->
                if (listener != null) {
                    val position = getAdapterPosition()
                    if (position != RecyclerView.NO_POSITION) {
                        listener!!.onItemClick(view1, position, "image")
                    }
                }
            }
            reduceQuantity.setOnClickListener { view1: View? ->
                if (listener != null) {
                    val position = getAdapterPosition()
                    if (position != RecyclerView.NO_POSITION) {
                        listener!!.onItemClick(view1, position, "reduceQuantity")
                    }
                }
            }
            increaseQuantity.setOnClickListener { view1: View? ->
                if (listener != null) {
                    val position = getAdapterPosition()
                    if (position != RecyclerView.NO_POSITION) {
                        listener!!.onItemClick(view1, position, "increaseQuantity")
                    }
                }
            }
        }
    }

    interface OnItemClickListener {
        fun onItemClick(view: View?, position: Int, type: String?)
    }
}
